[FEATURE] Support shebang in hooks

## Environment
- **OS**: Ubuntu (In ChromeOS)
- **Dotter version**: 0.12.4
- **Additional relevant information**

## Description
Hooks, like `post_deploy.sh`, are always executed in `sh`, ignoring any shebang line in the script.

## Reproduction
* Have a system where `sh` is not a symlink to `bash`.
* Create a `post_deploy.sh` file (or another hook) containing
  ```sh
  #!/usr/bin/env bash
  readlink /proc/$$/exe
  ```
* Run `dotter deploy` (or run the hook another way through Dotter).

### Expected behavior
The `readlink` command should show the shell listed in the shebang line (bash).

### Actual behavior
The `readlink` command shows `sh`, or in my case `dash`.

This is a problem because `sh` and `dash` don't have the same features as `bash` (or another script runner I might use). This makes my dotfiles less portable, as some systems use bash for the system shell (Arch), and some don't (Ubuntu).

Yes, I could write my script using only POSIX features, and it would probably run everywhere. However, right now I have a script that runs fine on one system, and breaks in confusing ways on another. The first system has `sh` symlinked to bash, and the second has it symlinked to dash.

The script in question is

```sh
#!/bin/bash
SYSTEMD_ROOT="${HOME}/.config/systemd/user"

function enable_service() {
  if [[ -f "${SYSTEMD_ROOT}/$1" ]] && ! [[ -f "${SYSTEMD_ROOT}/multi-user.target.wants/$1" ]]; then
    echo "Enabling and starting systemd unit $1"
    systemctl --user --now enable $1
  fi
}

{{ #if dotter.packages.taskwarrior }}
enable_service task-sync.timer
{{ /if }}
```

However, a `post_deploy.sh` as simple as this also fails:

```sh
#!/usr/bin/env bash
function foo() {
  echo hello from foo
}
foo
````

The error it gives is:

```
> dotter deploy -v
.dotter/cache/.dotter/post_deploy.sh: 2: .dotter/cache/.dotter/post_deploy.sh: Syntax error: "(" unexpected
[ERROR] Failed to deploy
Caused by:
    run post-deploy hook
    subshell returned error
```

It is certainly very surprising that the shebang line is ignored.

--- a/src/hooks.rs
+++ b/src/hooks.rs
@@ -5,8 +5,6 @@ use std::path::Path;
 use std::process::Command;
 use std::process::Child;
 
-use crate::config::TemplateTarget;
-
 pub(crate) fn run_hook(
     location: &Path,
     cache_dir: &Path,
@@ -19,16 +17,17 @@ pub(crate) fn run_hook(
     }
 
     let mut script_file = cache_dir.join(location);
+    let mut target = std::env::temp_dir().join("dotter_temp");
     if cfg!(windows) {
         script_file.set_extension("bat");
+        target.set_extension("bat");
     }
     debug!("Rendering script {:?} -> {:?}", location, script_file);
-
-    let target: TemplateTarget = std::env::temp_dir().join("dotter_temp").into();
+    
     crate::actions::perform_template_deploy(
         location,
         &script_file,
-        &target,
+        &target.clone().into(),
         &mut crate::filesystem::RealFilesystem::new(false),
         handlebars,
         variables,
@@ -36,13 +35,7 @@ pub(crate) fn run_hook(
     .context("deploy script")?;
 
     debug!("Running script file");
-    let mut child = if cfg!(windows) {
-        Command::new(target.target)
-            .spawn()
-            .context("spawn batch file")?
-    } else {
-        execute_unix(&target.target)?
-    };
+    let mut child = run_script_file(&target)?;
 
     anyhow::ensure!(
         child.wait().context("wait for child shell")?.success(),
@@ -53,7 +46,7 @@ pub(crate) fn run_hook(
 }
 
 #[cfg(unix)]
-fn execute_unix(script: &Path) -> Result<Child> {
+fn run_script_file(script: &Path) -> Result<Child> {
     use std::os::unix::fs::PermissionsExt;
 
     let permissions = script.metadata()?.permissions();
@@ -67,4 +60,11 @@ fn execute_unix(script: &Path) -> Result<Child> {
             .spawn()
             .context("spawn shell")
     }
+}
+
+#[cfg(windows)]
+fn run_script_file(script: &Path) -> Result<Child> {
+    Command::new(script)
+            .spawn()
+            .context("spawn batch file")
 }

--- a/src/hooks.rs
+++ b/src/hooks.rs
@@ -3,6 +3,7 @@ use handlebars::Handlebars;
 
 use std::path::Path;
 use std::process::Command;
+use std::process::Child;
 
 use crate::config::TemplateTarget;
 
@@ -40,19 +41,7 @@ pub(crate) fn run_hook(
             .spawn()
             .context("spawn batch file")?
     } else {
-        use std::os::unix::fs::PermissionsExt;
-        
-        let permissions = target.target.metadata()?.permissions();
-        if !target.target.is_dir() && permissions.mode() & 0o111 != 0 {
-            Command::new(target.target)
-                .spawn()
-                .context("spawn script file")?
-        } else {
-            Command::new("sh")
-                .arg(target.target)
-                .spawn()
-                .context("spawn shell")?
-        }
+        execute_unix(&target.target)?
     };
 
     anyhow::ensure!(
@@ -61,4 +50,21 @@ pub(crate) fn run_hook(
     );
 
     Ok(())
+}
+
+#[cfg(unix)]
+fn execute_unix(script: &Path) -> Result<Child> {
+    use std::os::unix::fs::PermissionsExt;
+
+    let permissions = script.metadata()?.permissions();
+    if !script.is_dir() && permissions.mode() & 0o111 != 0 {
+        Command::new(script)
+            .spawn()
+            .context("spawn script file")
+    } else {
+        Command::new("sh")
+            .arg(script)
+            .spawn()
+            .context("spawn shell")
+    }
 }

--- a/src/hooks.rs
+++ b/src/hooks.rs
@@ -4,6 +4,8 @@ use handlebars::Handlebars;
 use std::path::Path;
 use std::process::Command;
 
+use std::os::unix::fs::PermissionsExt;
+
 pub(crate) fn run_hook(
     location: &Path,
     cache_dir: &Path,
@@ -31,16 +33,23 @@ pub(crate) fn run_hook(
     )
     .context("deploy script")?;
 
-    debug!("Running script file ");
+    debug!("Running script file");
     let mut child = if cfg!(windows) {
         Command::new(script_file)
             .spawn()
             .context("spawn batch file")?
     } else {
-        Command::new("sh")
-            .arg(script_file)
-            .spawn()
-            .context("spawn shell")?
+        let permissions = script_file.metadata()?.permissions();
+        if !script_file.is_dir() && permissions.mode() & 0o111 != 0 {
+            Command::new(script_file)
+                .spawn()
+                .context("spawn script file")?
+        } else {
+            Command::new("sh")
+                .arg(script_file)
+                .spawn()
+                .context("spawn shell")?
+        }
     };
 
     anyhow::ensure!(

--- a/src/hooks.rs
+++ b/src/hooks.rs
@@ -2,8 +2,8 @@ use anyhow::{Context, Result};
 use handlebars::Handlebars;
 
 use std::path::Path;
-use std::process::Command;
 use std::process::Child;
+use std::process::Command;
 
 pub(crate) fn run_hook(
     location: &Path,
@@ -23,7 +23,7 @@ pub(crate) fn run_hook(
         target.set_extension("bat");
     }
     debug!("Rendering script {:?} -> {:?}", location, script_file);
-    
+
     crate::actions::perform_template_deploy(
         location,
         &script_file,
@@ -51,9 +51,7 @@ fn run_script_file(script: &Path) -> Result<Child> {
 
     let permissions = script.metadata()?.permissions();
     if !script.is_dir() && permissions.mode() & 0o111 != 0 {
-        Command::new(script)
-            .spawn()
-            .context("spawn script file")
+        Command::new(script).spawn().context("spawn script file")
     } else {
         Command::new("sh")
             .arg(script)
@@ -64,7 +62,5 @@ fn run_script_file(script: &Path) -> Result<Child> {
 
 #[cfg(windows)]
 fn run_script_file(script: &Path) -> Result<Child> {
-    Command::new(script)
-            .spawn()
-            .context("spawn batch file")
+    Command::new(script).spawn().context("spawn batch file")
 }

--- a/src/hooks.rs
+++ b/src/hooks.rs
@@ -4,7 +4,7 @@ use handlebars::Handlebars;
 use std::path::Path;
 use std::process::Command;
 
-use std::os::unix::fs::PermissionsExt;
+use crate::config::TemplateTarget;
 
 pub(crate) fn run_hook(
     location: &Path,
@@ -23,10 +23,11 @@ pub(crate) fn run_hook(
     }
     debug!("Rendering script {:?} -> {:?}", location, script_file);
 
+    let target: TemplateTarget = std::env::temp_dir().join("dotter_temp").into();
     crate::actions::perform_template_deploy(
         location,
         &script_file,
-        &std::env::temp_dir().join("dotter_temp").into(),
+        &target,
         &mut crate::filesystem::RealFilesystem::new(false),
         handlebars,
         variables,
@@ -35,18 +36,20 @@ pub(crate) fn run_hook(
 
     debug!("Running script file");
     let mut child = if cfg!(windows) {
-        Command::new(script_file)
+        Command::new(target.target)
             .spawn()
             .context("spawn batch file")?
     } else {
-        let permissions = script_file.metadata()?.permissions();
-        if !script_file.is_dir() && permissions.mode() & 0o111 != 0 {
-            Command::new(script_file)
+        use std::os::unix::fs::PermissionsExt;
+        
+        let permissions = target.target.metadata()?.permissions();
+        if !target.target.is_dir() && permissions.mode() & 0o111 != 0 {
+            Command::new(target.target)
                 .spawn()
                 .context("spawn script file")?
         } else {
             Command::new("sh")
-                .arg(script_file)
+                .arg(target.target)
                 .spawn()
                 .context("spawn shell")?
         }

